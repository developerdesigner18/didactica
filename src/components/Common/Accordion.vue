<template>
  <div :class="['accordion-wrapper', variant]">
    <div
      v-for="(item, index) in items"
      :key="index"
      :class="['accordion-item', { active: isOpen(index) }]"
    >
      <div class="accordion-header" @click="toggle(index)">
        <img
          src="https://whoalxppublicstorage.blob.core.windows.net/didactica/assets/images/awards/arrow.svg"
          loading="lazy"
        />
        <span>{{ item.title }}</span>
      </div>
      <div
        v-if="isOpen(index)"
        class="accordion-content"
        v-html="item.content"
      ></div>
    </div>
  </div>
</template>
<script>
export default {
  name: "Accordion",
  props: {
    items: {
      type: Array,
      required: true
    },
    singleOpen: {
      type: Boolean,
      default: true
    },
    variant: {
      type: String,
      default: "variant1"
    }
  },
  data() {
    return {
      openIndexes: []
    };
  },
  mounted() {
    if (!this.singleOpen && this.items.length > 0) {
      this.openIndexes.push(0);
    }
  },
  methods: {
    toggle(index) {
      if (this.singleOpen) {
        this.openIndexes = this.isOpen(index) ? [] : [index];
      } else {
        if (this.isOpen(index)) {
          this.openIndexes = this.openIndexes.filter((i) => i !== index);
        } else {
          this.openIndexes.push(index);
        }
      }
    },
    isOpen(index) {
      return this.openIndexes.includes(index);
    }
  }
};
</script>
<style lang="scss" scoped>
.accordion-wrapper {
  .accordion-item {
    &:not(:first-child) {
      margin-top: 16px;
    }
    &.active {
      .accordion-header {
        img {
          transform: rotate(0deg);
        }
      }
    }
  }
  .accordion-header {
    display: flex;
    align-items: center;
    overflow: hidden;
    @include label-large($font-color: $brand-neutral-900, $font-weight: 500);
    cursor: pointer;
    span {
      margin-left: 8px;
    }
    img {
      transform: rotate(90deg);
    }
  }
  .accordion-content {
    padding-left: 24px;
    @include label-medium;
    color: $brand-neutral-900;
    letter-spacing: 0.4px;
  }
}
</style>
