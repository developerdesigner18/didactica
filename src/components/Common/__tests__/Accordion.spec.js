import { createLocalVue, mount } from "@vue/test-utils";
import Accordion from "../Accordion.vue";
import { BImg } from "bootstrap-vue";

// create an extended `Vue` constructor
const localVue = createLocalVue();

localVue.component("accordion", Accordion);
localVue.component("b-img", BImg);

describe("Accordion.vue", () => {
  it("renders when created", async () => {
    let wrapper = await getWrapper();
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
    wrapper.destroy();
  });

  describe("Props", () => {
    let wrapper;

    afterEach(() => {
      wrapper.destroy();
    });

    it("should render all items passed as props", async () => {
      wrapper = await getWrapper();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll(".accordion-item");
      expect(items.length).toEqual(3);
    });

    it("should apply the correct variant class", async () => {
      wrapper = await getWrapper({ variant: "variant2" });
      await wrapper.vm.$nextTick();

      expect(wrapper.classes()).toContain("variant2");
    });
  });

  describe("Functionality", () => {
    let wrapper;

    beforeEach(async () => {
      wrapper = await getWrapper();
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it("should toggle item content on click", async () => {
      const firstItemHeader = wrapper.find(".accordion-item .accordion-header");
      await firstItemHeader.trigger("click");
      await wrapper.vm.$nextTick();

      const firstItemContent = wrapper.find(
        ".accordion-item .accordion-content"
      );
      expect(firstItemContent.isVisible()).toBeTruthy();

      await firstItemHeader.trigger("click");
      await wrapper.vm.$nextTick();

      expect(
        wrapper.find(".accordion-item .accordion-content").exists()
      ).toBeFalsy();
    });
  });
});

function getWrapper(props = {}) {
  return mount(Accordion, {
    propsData: {
      items: [
        { title: "Item 1", content: "Content 1" },
        { title: "Item 2", content: "Content 2" },
        { title: "Item 3", content: "Content 3" }
      ],
      ...props
    },
    localVue
  });
}
