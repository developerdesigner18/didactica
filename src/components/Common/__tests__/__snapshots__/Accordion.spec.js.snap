// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion.vue renders when created 1`] = `
<div
  class="accordion-wrapper variant1"
>
  <div
    class="accordion-item"
  >
    <div
      class="accordion-header"
    >
      <img
        loading="lazy"
        src="https://whoalxppublicstorage.blob.core.windows.net/didactica/assets/images/awards/arrow.svg"
      />
       
      <span>
        Item 1
      </span>
    </div>
     
    <!---->
  </div>
  <div
    class="accordion-item"
  >
    <div
      class="accordion-header"
    >
      <img
        loading="lazy"
        src="https://whoalxppublicstorage.blob.core.windows.net/didactica/assets/images/awards/arrow.svg"
      />
       
      <span>
        Item 2
      </span>
    </div>
     
    <!---->
  </div>
  <div
    class="accordion-item"
  >
    <div
      class="accordion-header"
    >
      <img
        loading="lazy"
        src="https://whoalxppublicstorage.blob.core.windows.net/didactica/assets/images/awards/arrow.svg"
      />
       
      <span>
        Item 3
      </span>
    </div>
     
    <!---->
  </div>
</div>
`;
